dartagnan/src/main/java/com/dat3m/dartagnan/program/event/visitors/CompilationVisitor.java

```diff
@@ -0,0 +1,860 @@
+package com.dat3m.dartagnan.program.event.visitors;
+
+import com.dat3m.dartagnan.configuration.Arch;
+import com.dat3m.dartagnan.expression.*;
+import com.dat3m.dartagnan.expression.op.*;
+import com.dat3m.dartagnan.program.Register;
+import com.dat3m.dartagnan.program.event.arch.tso.*;
+import com.dat3m.dartagnan.program.event.Tag;
+import com.dat3m.dartagnan.program.event.Tag.C11;
+import com.dat3m.dartagnan.program.event.arch.aarch64.*;
+import com.dat3m.dartagnan.program.event.core.*;
+import com.dat3m.dartagnan.program.event.core.rmw.*;
+import com.dat3m.dartagnan.program.event.lang.catomic.*;
+import com.dat3m.dartagnan.program.event.lang.linux.*;
+import com.dat3m.dartagnan.program.event.lang.linux.cond.*;
+import com.dat3m.dartagnan.program.event.lang.pthread.*;
+
+import static com.dat3m.dartagnan.program.event.EventFactory.*;
+import static com.dat3m.dartagnan.program.event.EventFactory.AArch64;
+import static com.dat3m.dartagnan.program.event.EventFactory.Linux;
+import static com.dat3m.dartagnan.program.event.Tag.*;
+
+import com.google.common.base.Preconditions;
+
+import static com.dat3m.dartagnan.configuration.Arch.POWER;
+import static com.dat3m.dartagnan.expression.op.COpBin.EQ;
+import static com.dat3m.dartagnan.expression.op.COpBin.NEQ;
+
+import java.math.BigInteger;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+public class CompilationVisitor implements EventVisitor<List<Event>> {
+
+	Arch target;
+	
+	public CompilationVisitor(Arch target) {
+		this.target = target;
+	}
```

Comment:
You can make the `target` a `private final` field. Also, check for `null` only once on construction and get rid of all the unnecessary `Precondition.notNull` checks in the visit methods.